src/lower/wf.rs

```diff
@@ -0,0 +1,272 @@
+use std::sync::Arc;
+
+use ir::*;
+use errors::*;
+use cast::Cast;
+use solve::SolverChoice;
+use itertools::Itertools;
+
+struct WfSolver {
+    env: Arc<ProgramEnvironment>,
+    solver_choice: SolverChoice,
+}
+
+impl Program {
+    pub fn verify_well_formedness(&self, solver_choice: SolverChoice) -> Result<()> {
+        set_current_program(&Arc::new(self.clone()), || self.solve_wf_requirements(solver_choice))
+    }
+
+    fn solve_wf_requirements(&self, solver_choice: SolverChoice) -> Result<()> {
+        let solver = WfSolver {
+            env: Arc::new(self.environment()),
+            solver_choice,
+        };
+
+        for (id, struct_datum) in &self.struct_data {
+            if !solver.verify_struct_decl(struct_datum) {
+                let name = self.type_kinds.get(id).unwrap().name;
+                return Err(Error::from_kind(ErrorKind::IllFormedTypeDecl(name)));
+            }
+        }
+
+        for impl_datum in self.impl_data.values() {
+            if !solver.verify_trait_impl(impl_datum) {
+                let trait_ref = impl_datum.binders.value.trait_ref.trait_ref();
+                let name = self.type_kinds.get(&trait_ref.trait_id).unwrap().name;
+                return Err(Error::from_kind(ErrorKind::IllFormedTraitImpl(name)));
+            }
+        }
+
+        Ok(())
+    }
+}
+
+/// A trait for retrieving all types appearing in some Chalk construction.
+trait FoldInputTypes {
+    fn fold(&self, accumulator: &mut Vec<Ty>);
+}
+
+impl<T: FoldInputTypes> FoldInputTypes for Vec<T> {
+    fn fold(&self, accumulator: &mut Vec<Ty>) {
+        for f in self {
+            f.fold(accumulator);
+        }
+    }
+}
+
+impl FoldInputTypes for Parameter {
+    fn fold(&self, accumulator: &mut Vec<Ty>) {
+        match *self {
+            ParameterKind::Ty(ref ty) => ty.fold(accumulator),
+            _ => (),
+        }
+    }
+}
+
+impl FoldInputTypes for Ty {
+    fn fold(&self, accumulator: &mut Vec<Ty>) {
+        match *self {
+            Ty::Apply(ref app) => {
+                accumulator.push(self.clone());
+                app.parameters.fold(accumulator);
+            }
+            Ty::Projection(ref proj) => {
+                accumulator.push(self.clone());
+                proj.parameters.fold(accumulator);
+            }
+            Ty::UnselectedProjection(ref proj) => {
+                accumulator.push(self.clone());
+                proj.parameters.fold(accumulator);
+            }
+
+            // Type parameters and higher-kinded types do not carry any input types (so we can sort
+            // of assume they are always WF).
+            Ty::Var(..) | Ty::ForAll(..) => (),
```

Comment:
Hmm, `Ty::ForAll` corresponds to `for<'a> fn(&'a u32)` or some such. I don't know that we can assume they are always WF. I think this accumulator probably doesn't want to accumulate into a vector, but rather to invoke a callback, because in the case of `for<'a>` we may want to create subuniverses? I have to read this code a bit more carefully I guess.

Reply:
Actually we discussed that and for some reason we decided that higher-kinded types are always WF. This is similar to the fact that if we have this kind of function:
```rust
fn foo<T>(arg: HashSet<NotHash<T>>) { ... }
```
we would not issue any error until the function is called. For lifetime parametric function types, again an error would be issued only if a function of this type is ever called. We discussed that in our dropbox paper (see 2017.07.25).

Reply:
Oh, right, I forgot about that. Basically we were assuming that for all types introduce their *own* implied bounds, right? Makes sense.

Reply:
Can we put a more explicit comment then? =)